Toolbar Integration

timbittner opened the pull request:
Closes #22
Closes #23 
Closes #26 
Closes #27 

Implements FX Toolbar (UI) with ProjectExplorer related buttons.
Also provides placeholders for future command support (#34)
Implements Search Filtering on RelativePath.contains(str)
Implements View Representation Change (Flat vs Hierarchical)
Lastly changes the Icons for Projects and Folders.

Commit:
Add new icons

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/27

Commit:
Update folder and project icons

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/27

Commit:
Add build flat tree

- AbstractFileTreeElement now knows if it should display the whole
relative path or only the file name (also added this to the interface)
- CustomTreeCell now always shows the file name when editing
- Add functionality to build a flat tree to the
ProjectExplorerViewController

Task-Url: https://github.com/Jooshua/e4CompareFramework/issues/23

timbittner committed:
Add Toolbar UI for Search and Commands.

timbittner committed:
Add ProjecExplorer related tool bar buttons.

Task #26

timbittner committed:
Add Keycombos NewFolder and ShowInExplorer

Also update documentation.

Task #26

timbittner committed:
Merge branch 'feature/task_28-multiselection' into
feature/task_22-search

timbittner committed:
Add basic filtering to tree view via node's relative path match.

Task #22

timbittner committed:
Merge remote-tracking branch
'origin/feature/task_23-Add_flat_view_of_directories' into
feature/task_22-search

timbittner committed:
Replace expansion state mechanisms with listeners on TreeItems.

#22

timbittner committed:
Merge remote-tracking branch 'origin/feature/task_27-New_icons' into feature/task_22-search

timbittner committed:
Add Icons for View Representation.

Also update context menu icon for new folder.

#22

valentinpurrucker commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/controller/ProjectExplorerViewController.java:
```diff
@@ -118,29 +129,92 @@ public void initializeView(IEclipseContext context, WorkspaceFileSystem fileSyst
+		TreeItem<FileTreeElement> currentNode = new TreeItem<FileTreeElement>(parentNode, imgNode);
+
+		// Update our tree expansion state when a node's expansion state changes
+		currentNode.expandedProperty().addListener(new ChangeListener<Boolean>() {
```
Maybe convert to lambda.

valentinpurrucker commented on de.tu_bs.cs.isf.e4cf.core/src/de/tu_bs/cs/isf/e4cf/core/util/services/RCPImageService.java:
```diff
@@ -86,6 +90,19 @@ public Image getImage(String bundleName, String imagePath) {
+	 * This method returns a FX ImageView (Node) from his description it uses {@link getImageDescriptor()}.
+	 * @param bundleName the bundle name of the image.
+	 * @param imagePath the path within the bundle.
+	 * @param imageFileName the name of the image with extension.
```
imageFileName param does not exist in the method declaration.

valentinpurrucker commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/controller/ProjectExplorerToolBarController.java:
```diff
@@ -0,0 +1,194 @@
+		// Search Section
+		TextField search = new TextField();
+		search.setVisible(false);
+		search.setOnAction(actionEvent -> {
+			services.eventBroker.send(E4CEventTable.EVENT_FILTER_CHANGED, search.getText());
+		});
```
Would be cool if the view updates in real time with every keystroke. But maybe thats too demanding to update all the time.

timbittner replied:
As we need to update the tree every time I would hesitate for now.
I could imagine larger workspaces may be heavily impacted by such behaviour.

If no one else speaks out about it, I would postpone that and maybe make another issue for that feature.
Maybe including a switch to toggle the behaviour.

valentinpurrucker commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/controller/ProjectExplorerToolBarController.java:
```diff
@@ -0,0 +1,194 @@
+		searchButton.setOnAction(actionEvent -> {
+			search.setVisible(!search.isVisible());
+			// Removing and readding the search bar allows automatic resizing
+			if (search.isVisible()) {
+				searchButton.setGraphic(xImage);
+				searchButton.setTooltip(closeSearchTooltip);
+				bar.getItems().add(1, search);
```
We could add that as soon as the searchbar(textfield) is visible it also gets the focus so the user can directly start typing.

valentinpurrucker reviewed (commented):
Looks really good and as far as i can say, it works. I only found some small things.

Jooshua commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/project_explorer.e4xmi:
```diff
@@ -4,7 +4,7 @@
     <elements xsi:type="basic:Part" xmi:id="_RuDIk_foEeiagocgPMLR0w" elementId="de.tu_bs.cs.isf.e4cf.parts.project_explorer" contributionURI="bundleclass://de.tu_bs.cs.isf.e4cf.parts.project_explorer/de.tu_bs.cs.isf.e4cf.parts.project_explorer.controller.ProjectExplorerViewAdapter" label="Project Explorer" iconURI="platform:/plugin/de.tu_bs.cs.isf.e4cf.core/icons/Explorer_View/items/project_explo ...
       <menus xsi:type="menu:PopupMenu" xmi:id="_RuDIlffoEeiagocgPMLR0w" elementId="de.tu_bs.cs.isf.e4cf.parts.project_explorer.context_menu" label="project_explorer_context_menu">
-        <children xsi:type="menu:DirectMenuItem" xmi:id="_RuDIlvfoEeiagocgPMLR0w" elementId="de.tu_bs.cs.isf.familymining.ppu_iec.rcp_e4.directmenuitem.new_folder" label="New Folder" iconURI="platform:/plugin/de.tu_bs.cs.isf.e4cf.core/icons/Explorer_View/items/project16.png" contributionURI="bundleclass://de.tu_bs.cs.isf.e4cf.parts.project_explorer/de.tu_bs.cs.isf.e4cf.parts.project_explorer.hand ...
+        <children xsi:type="menu:DirectMenuItem" xmi:id="_RuDIlvfoEeiagocgPMLR0w" elementId="de.tu_bs.cs.isf.familymining.ppu_iec.rcp_e4.directmenuitem.new_folder" label="New Folder" iconURI="platform:/plugin/de.tu_bs.cs.isf.e4cf.core/icons/Explorer_View/items/folder16.png" contributionURI="bundleclass://de.tu_bs.cs.isf.e4cf.parts.project_explorer/de.tu_bs.cs.isf.e4cf.parts.project_explorer.handl ...
```
Icon inconsistency: the new folder wizard uses blue icon (project) but toolbar and now context menu that call the wizard use the yellow icon (directory)

Jooshua commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/controller/ProjectExplorerToolBarController.java:
```diff
@@ -0,0 +1,194 @@
+	private Button btnImportFiles;
+	private Button btnDelete;
+	private Button btnShowExplorer;
+	boolean isFlatView = false;
```
private? new line before?

Jooshua commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/controller/ProjectExplorerToolBarController.java:
```diff
@@ -0,0 +1,194 @@
+
+		ImageView searchImage = services.imageService.getFXImage(null, FileTable.SEARCH_PNG);
+		Tooltip searchTooltip = new Tooltip("Search Files");
+		ImageView xImage = services.imageService.getFXImage(null, FileTable.X_PNG);
```
Maybe a more spelling name, e.g. `closeSearchImage` ?

Jooshua commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/controller/ProjectExplorerToolBarController.java:
```diff
@@ -0,0 +1,194 @@
+		searchButton.setTooltip(searchTooltip);
+		searchButton.setOnAction(actionEvent -> {
+			search.setVisible(!search.isVisible());
+			// Removing and readding the search bar allows automatic resizing
```
```suggestion
			// Removing and reading the search bar allows automatic resizing
```

timbittner replied:
Its re-adding not reading like in read a book

Jooshua replied:
Then lets write re-adding ^^

Jooshua commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/controller/ProjectExplorerToolBarController.java:
```diff
@@ -0,0 +1,194 @@
+		Button searchButton = new Button("", searchImage);
+		searchButton.setTooltip(searchTooltip);
+		searchButton.setOnAction(actionEvent -> {
+			search.setVisible(!search.isVisible());
```
Why do we set the visibility when we just add if visible or remove if unvisible in the end?

timbittner replied:
Visibility changes the rendering of a component. It leaves the allocated space in the toolbar behind, resulting in empty gray fields.

Sometimes listeners and other subscribers change their behaviour on the visibility/activeness of a component. To be safe I toggle it.

Jooshua commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/controller/ProjectExplorerToolBarController.java:
```diff
@@ -0,0 +1,194 @@
+		});
+
+		// Delete
+		btnDelete = createToolbarButton("Delete Selected Files", FileTable.DELETE_PNG, actionEvent -> {
```
Toolbar text is misleading: We can also use this btn to delete directories

Jooshua commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/controller/ProjectExplorerToolBarController.java:
```diff
@@ -0,0 +1,194 @@
+
+		// Delete
+		btnDelete = createToolbarButton("Delete Selected Files", FileTable.DELETE_PNG, actionEvent -> {
+			RemoveFileCommand h = new RemoveFileCommand();
```
`h` is not a good name

Jooshua commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/controller/ProjectExplorerToolBarController.java:
```diff
@@ -0,0 +1,194 @@
+			}
+		});
+		bar.getItems().add(btnToggleView);
+		
+
```
Remove nl

Jooshua commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/controller/ProjectExplorerViewController.java:
```diff
@@ -118,29 +129,92 @@ public void initializeView(IEclipseContext context, WorkspaceFileSystem fileSyst
 		});
 
+		// Handoff Toolbar
+		toolbarController = new ProjectExplorerToolBarController(projectToolbar, services,
+				canvas.getParent().getShell());
```
We can open the context menu with right clicking on the toolbar (because we register it on the canvas here).
Maybe we can only use it on the tree with adding a subcanvas or something like that.

timbittner replied:
I tested that and attaching the context menu directly onto the project tree object but without success.
The behavior is weird but not application breaking so I propose ignoring it.

Jooshua replied:
Yeah I thought about it again. I think the only way is we possible could resolve this is using two SWT canvas for each the toolbar and the tree that would have to be split in different fxml files, which kinda sucks. So lets leave it for now...

Jooshua commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/stringtable/FileTable.java:
```diff
@@ -3,12 +3,24 @@
+			.toString();
+
 	// Image paths
 	public static String PROJECT_EXPLORER_ICONS = Paths.get("icons", "Explorer_View", "items").toString();
```
Add a constant for "icons", "Explorer_View"

Jooshua reviewed (changes requested):
Overall nice, but some minor things we have to change / discuss.

timbittner committed:
Change Requests from Pull #35

Update NewFolderWizard image
Update JavaDoc
Make search field request focus on appear
Minor improvements

#35

timbittner committed:
Remove Placeholders for Undo/Redo

#35
